sklearn/impute/_base.py

```diff
@@ -395,32 +412,13 @@ def fit(self, X, y=None):
         else:
             fill_value = self.fill_value
 
-        # fill_value should be numerical in case of numerical input
-        if (
-            self.strategy == "constant"
-            and X.dtype.kind in ("i", "u", "f")
-            and not isinstance(fill_value, numbers.Real)
-        ):
-            raise ValueError(
-                "'fill_value'={0} is invalid. Expected a "
-                "numerical value when imputing numerical "
-                "data".format(fill_value)
-            )
+        if self.strategy == "constant":
+            _check_fill_values_dtype([fill_value], X.dtype)
```

Comment:
I think that we don't need the function `_check_fill_values_dtype` and instead make the check in the `_validate_input`. The reason is that `_validate_input` already have a `in_fit=True/False` so we can either validate `self.fill_values` or use `self.statistics_`. I assume that we can return those since they are required in the rest of fit or transform.